### ISSUE TITLE
Issues with Html.Hidden

### ISSUE THREAD

**Issue by EdwardFinn** [aspnet/AspNetWebStack#258]
### Describe the bug
Html.Hidden does not consistently render with the value passed in. May require binding to a collection that was mutated.

### To Reproduce
1. Using ASP.NET 4.6.2, Mvc 5.2.7, Razor 3.2.7
2. I've attached a reproduction project
[WebApplication1.zip](https://github.com/aspnet/AspNetWebStack/files/3628302/WebApplication1.zip)
Note: It may require this nuget command to run first time.
   - Update-Package Microsoft.CodeDom.Providers.DotNetCompilerPlatform -r

### Expected behavior
Expected the value of the string property to be the same inside the input rendered by Html.Hidden as it does when output as literal text

### Screenshots
From the Example Application
Note: Hidden inputs have been forced to type=text with JavaScript after server response.

Start with a 'ViewModel' containing a collection of 'Models' with only a TimeSpan property. Initialize the collection with only one model containing  value 12:00:00. Litteral, Html.Hidden and Input of type hidden all show the same value
![RazorHiddenBug01](https://user-images.githubusercontent.com/11775457/65184885-7229b000-da34-11e9-9785-7c7e79c6852c.png)

Click button 'Step 1' this sends, via ajax, a serialized form that binds on the hidden field(s), and add TimeSpan with value of 10:00:00 to the ViewModel collection, server responses with partial view and html is replaced. Values all show as expected
![RazorHiddenBug02](https://user-images.githubusercontent.com/11775457/65184889-748c0a00-da34-11e9-9560-bd1fbd73014f.png)

Click button 'Step 2' this sends, via ajax, a serialized form that binds on the hidden field(s), and deletes the model containing TimeSpan with value of 12:00:00 from the ViewModel collection. **This is the bug.** All three fields are showing the same property, but the hidden input generated by Html.Hidden now displays the wrong value '12:00:00
![RazorHiddenBug03](https://user-images.githubusercontent.com/11775457/65184891-75bd3700-da34-11e9-9101-0177dcb4dc91.png)

Click button 'Step 3' and the wrong value from the hidden field now binds and displays the erroneous value to the user
![RazorHiddenBug04](https://user-images.githubusercontent.com/11775457/65184905-78b82780-da34-11e9-999a-8fa4a55309a6.png)

**Comment by RichardD2** [comment]
I ran into something similar a few weeks back. On reflection, I think it's the expected behaviour most of the time.

When you use an HTML helper to render an input, it will try to take the value submitted by the user in preference to the value from the model. This is to allow for cases where the user has entered a value which cannot be converted to the model property's type - for example, entering "forty two" for an `int` property. The idea being that it's better to show the user what they typed and let them correct it, rather than have their input disappear completely.

In step 2 of your example, you remove the item at index 0 (`12:00:00`). That means your model now only contains one item, which is now at index 0 (`10:00:00`). When you render the hidden input for `exampleModels[0].timeSpan`, the helper method sees that the user submitted a value for that field (`12:00:00`), and outputs that value instead of the value from the model.

The literal output and the vanilla hidden input don't include this logic, so they simply output the value from the model (`10:00:00`).

In step 3, you submit the updated value of `12:00:00`, which updates the model property. Now all three fields show the same value, which is the value you submitted, although not the value you were expecting.

You can demonstrate this behaviour in a single step, without requiring a collection or an AJAX request.
```
class ExampleViewModel
{
    public TimeSpan timeSpan { get; set; }
    public string timeSpanString => timeSpan.ToString();
}
```

```
class HomeController : Controller
{
    [HttpGet]
    public ActionResult Index() => View(new ExampleViewModel 
    { 
        timeSpan = TimeSpan.FromHours(12) 
    });
    
    [HttpPost]
    public ActionResult Index(ExampleViewModel model) => View(model);
}
```

```
@model ExampleViewModel

@using (Html.BeginForm())
{
    <p>
        The string value (@Model.timeSpanString) <br/>
        The Razor value ( @Html.TextBox("timeSpan", Model.timeSpanString) ) <br/>
        The vanilla value ( <input id="raw.timeSpan"  type="text" value="@Model.timeSpanString" disabled /> ) <br/>
    </p>

    <p><input type="submit" /></p>
}
```

* Run the application - all three values show "12:00:00".
* Type "test" into the Razor input, and submit the form;
* The static and vanilla values now show "00:00:00", and the Razor input shows "test";

**Comment by EdwardFinn** [comment]
I agree the example isn't the simplest form to reproduce, but it was how to code I debugged was written and I wanted to be sure it was reproducible. 

But I also disagree with you take on the reason, as I actually ran into this issue again this week with my own code and gained a better realization. The problem revolves around how Html.Hidden uses ModelState. 

If you take my example application, and add ModelState.Clear() right before calling Partial view.
`
ModelState.Clear();
return PartialView("_contentPartial", vModel);
`

All the sudden everything works as expected. 

So I did more digging, if you debug ModelState while the result of Step2 is rendering. You can see the ModelState contains the data send from the browser not reflecting our mutated state. So while our new model has only 1 record at index 0 with a value of 10:00, our ModelState has our bound value of 12:00 at index 0, and 10:00 at it's old index 1.
![hiddenModelStateIssue](https://user-images.githubusercontent.com/11775457/67445307-72404100-f5da-11e9-9de9-85556bc55038.png)

This means Html.Hidden is rendering the value from ModelState, not the actual variable passed to it.

I assert this should not be intended functionality.

**Comment by dougbu** [comment]
@EdwardFinn what @RichardD2 said is correct. The `ModelState` data contains data the user submitted, falling back to the current `Model` only when such data is unavailable. And, HTML generation uses what's in `ModelState` in most cases.  This is, as @RichardD2 said the intended behaviour to correctly round-trip erroneous user entries.

**Comment by EdwardFinn** [comment]
@dougbu I see, in my eagerness to detail my newest findings I seem to have read, yet not comprehended the response from @RichardD2 . My apologies for that RichardD2.

On a second reading I see we have both detailed the same underlying action in our own words, and your reasons make sense..

I expected ModelState to only persist the Error for validation controls. But if we take RichardD2's Scenario, we can see why it works this way.
 
If my model contained an int, but a user entered text 'forty two'. On model binding, my model has a int value of 0, since 'forty two' can't be parsed to an int. But ModelState will know this and mark it as not valid while retaining the user entered value (as string). Without pulling from ModelState, rebinding the model would destroy the user input and display the value '0'. So the HTML helpers target the ModelState first to give the user their entered data 'forty two' and the validation message it must be numeric.

In other words there is no bug here. The functionality is un-intuitive when the bound collection is altered server side, but otherwise a true feature to enhance user experience.

The solution to continue utilizing the HTML helpers with this methodology, would probably be to only change the collection if the ModelState.iSValid returns true, then clear ModelState after.

Thank You @dougbu & @RichardD2 for your time explaining this.
I understand now this is not a bug, and the ticket can be closed.
